File: app/src/main/java/com/example/budgetbuddy/database/CategoryDAO.java
package com.example.budgetbuddy.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;

import com.example.budgetbuddy.entities.Category;

import java.util.List;


/**
 * Data Access Object for the Category Entity. Defines CRUD methods
 */
@Dao
public interface CategoryDAO
{
    /**
     * Get all categories
     * @return a <code>LiveData</code> object containing a list of <code>Category</code> objects
     */
    @Query("SELECT * FROM 'category'")
    LiveData<List<Category>> getAll();

    /**
     * Get a category by its ID
     * @param id the ID of the category
     * @return a <code>Category</code> object
     */
    @Query("SELECT * FROM 'category' WHERE categoryID = (:id)")
    Category getCategoryByID(long id);

    /**
     * Insert a new category into the database
     * @param categories the <code>Category</code> object to insert
     */
    @Insert
    void insertCategory(Category... categories);

    /**
     * Delete a category from the database
     * @param category the <code>Category</code> object to delete
     */
    @Delete
    void delete(Category category);


}

File: app/src/main/java/com/example/budgetbuddy/data/db/CategoryDao.kt
package com.example.budgetbuddy.data.db

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import com.example.budgetbuddy.data.entities.Category
import kotlinx.coroutines.flow.Flow

@Dao
interface CategoryDao
{
    /**
     * Inserts any number of Category objects into the database.
     */
    @Insert
    suspend fun insertAll(vararg categories: Category)

    /**
     * Deletes a Category object from the database
     */
    @Delete
    suspend fun delete(category: Category)

    /**
     * Returns all Category objects stored in the database as a Flow list.
     * Ordered by name in ascending order by default
     */
    @Query("SELECT * FROM 'category' ORDER BY 'name'")
    fun getAll(): Flow<List<Category>>

    /**
     * Updates a category located by ID. Existing category data is replaced
     * with data extracted from provided Category object.
     */
    @Update
    suspend fun update(category: Category)
}